### PR TITLE
Merlin server configuration through commands

### PR DESCRIPTION
**Pull request by ryannova**
Current draft of the merlin server config command with a list of made changes and a list of up and coming changes and tested functionalities.

Changes made:
- Create functional calls for each merlin server command
- Create server_command.py to store command calls.
- Added config subcommand to merlin server
- Added following flags to config subcommand
  - ipaddress (Set the binded ip address of the container)
  - port (Set the binded port of the container)
  - user (Set the main user file for container)
  - password (Set the main user password file for container)
  - add-user (Add a user to the container image [outputs an associated password file for user]) (ONLY CHANGE IN CONFIG FILE)
  - remote-user (Remove user from list of added users) (ONLY CHANGE IN CONFIG FILE)
  - directory (Set the directory of the merlin server container files)
  - snapshot-seconds (Set the number of seconds elapsed before snapshot change condition is checked)
  - snapshot-changes (Set snapshot change condition for a snapshot to be made)
  - snapshot-file (Set the database file that the snapshot will be written to)
  - append-mode (Set the append mode for redis)
  - append-file (Set the name of the append only file for redis)
- Added user_file to merlin server config (user file lives in local config directory and list users and their associated permissions)
- Added pass_file to merlin server config (pass file lives in main config directory and is the main password used for all merlin server instances)
- Enable masteruser in redis config
- Enable requirepass in redis config 
- Moved default constants from server_setup.py to server_config.py
- Added server_util.py file to store utility functions.
- Created RedisConfig class to interface with redis.conf file
- Created RedisUsers class to interface with redis.user file

Changes still need to be made:
- add_user function needs to add user to exisiting merlin server instance if one is running
- remove_user function needs to remove user from merlin server instance if one is running
- (Potential change) Better interface with getting configuration values from merlin server config file. Currently requires and long one line if statement

Tested:
- Authenticating through redis-cli by providing file with password.

Things to test:
- Test on LC machines.
- Test using app.yaml main configuration files
- Test with merlin workflows
- Test any missing edge cases

### CODE REVIEW

**Comment by koning** [comment]
What is the process to test this on LC?

**Comment by koning** [comment]
```
(merlin3_8) [host]:merlin server init  
[2022-06-09 15:02:13: INFO] Unable to find exisiting server configuration.
[2022-06-09 15:02:13: INFO] Creating default configuration in /<path>/.merlin/server/
[2022-06-09 15:02:13: INFO] Copying file singularity.yaml to configuration directory.
[2022-06-09 15:02:13: INFO] Copying file docker.yaml to configuration directory.
[2022-06-09 15:02:13: INFO] Copying file podman.yaml to configuration directory.
[2022-06-09 15:02:13: INFO] Copying file merlin_server.yaml to configuration directory.
[2022-06-09 15:02:13: INFO] Creating merlin server directory.
[2022-06-09 15:02:13: INFO] Creating password file for merlin server container.
[2022-06-09 15:02:13: INFO] User <user> created in user file for merlin server container
[2022-06-09 15:02:13: INFO] Fetching redis image from docker://redis
[2022-06-09 15:02:26: INFO] Copying default redis configuration file.
[2022-06-09 15:02:26: INFO] Merlin server initialization successful.
(merlin3_8) [host]:merlin server start
[2022-06-09 15:02:37: ERROR] cannot unpack non-iterable NoneType object
```

**Comment by ryannova** [comment]
I made some slight changes to the repo when I was testing this on the LC. Try to pull the lastest version and let me know if it works.

As for testing on the LC here are some tests that you can try.

Merlin Server basic functionality test
1. merlin server init
2. merlin server start
4. merlin server info
5. redis-cli
6. AUTH merlin_password
7. ACL list
8. exit
9. merlin server stop

Merlin Server restart functionality test
1. merlin server init
2. merlin server start
3. merlin server restart
4. merlin server info
5. redis-cli
6. AUTH merlin_password
7. ACL list
8. exit
9. merlin server stop

Note here that you can AUTH as your current user by using your current username on the machine and the password located in "merlin_server/redis.pass". You will need to open the file can copy the random string and use it in step 6 (AUTH <user> <password>)

Merlin workflow test
1. merlin server init
2. merlin server start
3. merlin server info
4. merlin example hello_samples
5. merlin run hello/hello_samples.yaml
6. merlin server stop

Somethings to note about your merlin config file in "~/.merlin/app.yaml". Make sure that you have either "localhost" or "127.0.0.1" set as the server and the user set to "default" and password set to "merlin_password" (For some reason its not able to read a redis.pass file which contains the password for the current user. I'm still looking into it.)

I'm also working on putting something together that I can show working on LC and hopefully in another pull request I can make some changes to the current testing framework to allow for automating some of these tests.

**Comment by ryannova** [comment]
I've updated the changelogs to reflect the changes made in this pull request. @lucpeterson @koning Please take a look at it and if everything looks good, it is ready to be merged into develop.

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,11 +14,39 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 - Added documentation page docs/merlin_server.rst, docs/modules/server/configuration.rst, and docs/modules/server/commands.rst
+- Added merlin server config command for editing configuration files.
+- Added server_command.py to store command calls.
+- Added following flags to config subcommand
+  - ipaddress (Set the binded ip address of the container)
+  - port (Set the binded port of the container)
+  - user (Set the main user file for container)
+  - password (Set the main user password file for container)
+  - add-user (Add a user to the container image [outputs an associated password file for user])
+  - remote-user (Remove user from list of added users)
```
Comment (lucpeterson) [inline CHANGELOG.md:25]: i think "remote-user" should be "remove-user"

File: merlin/main.py [hunk 2]
```diff
@@ -626,6 +609,101 @@ def setup_argparse() -> None:
...
+    server_config.add_argument(
+        "-pwd",
+        "--password",
+        action="store",
+        type=str,
+        # default="~/.merlin/redis.pass",
+        help="Set the password file to be used for merlin server container.",
+    )
+    server_config.add_argument(
+        "--add-user",
```
Comment (lucpeterson) [inline merlin/main.py:651]: is there a "remove-user" command? I can't find it

**Review by lucpeterson — APPROVED** [review]
a few pending comments. but otherwise looks good!

File: merlin/main.py (hunk 2 above)
Comment (ryannova) [reply merlin/main.py:651]: There are add and remove user flags to the config command.
